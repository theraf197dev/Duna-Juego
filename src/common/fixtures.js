import trialImage from './media/trial_mode.jpg';
import seekerImage from './media/seeker_mode.jpg';

const controls = [
    {
        id: 1,
        name: 'Up',
        vector: 'x',
        incr: -1,
    },
    {
        id: 2,
        name: 'Down',
        vector: 'x',
        incr: 1,
    },
    {
        id: 3,
        name: 'Left',
        vector: 'y',
        incr: -1,
    },
    {
        id: 4,
        name: 'Right',
        vector: 'y',
        incr: 1,
    },
];

const games = {
    items: [
        {
            name: 'TRIAL',
            url: trialImage,
        },
        {
            name: 'SEEKER',
            url: seekerImage,
        },
    ],
};

const difficulty = 1;
const mode = 'TRIAL';
const size = 9;

export {
    controls,
    difficulty,
    games,
    mode,
    size,
};
