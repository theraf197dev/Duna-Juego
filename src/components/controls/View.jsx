import React from 'react';
import {
  ButtonStyles,
  CircleStyles,
  ContainerStyles,
  CrossStyles,
  FigureStyles,
} from './View.styles.jsx';
import { canMove, GameStatus } from '../../common/utils.js';

const View = ({
    controls,
    currentPos,
    gameStatus,
    handleBark,
    handleChangePosition,
    matrix,
    size,
}) => {
  return (
    <ContainerStyles>
      {controls.map((control) =>
        <ButtonStyles
          control={control}
          data-testid={`fn-controls-button-${control.name}`}
          disabled={
            !canMove(matrix, currentPos, control.vector, control.incr, size)
            || gameStatus !== GameStatus.playing
          }
          key={control.id}
          onClick={() => handleChangePosition(control.vector, control.incr)}
        >
          <FigureStyles control={control} />
        </ButtonStyles>
      )}
      <CrossStyles>
        <CircleStyles data-testid={'fn-controls-bark'} onClick={handleBark} />
      </CrossStyles>
    </ContainerStyles>
  )
}

export default View