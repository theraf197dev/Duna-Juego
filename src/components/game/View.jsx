import React from 'react';
import {
    CellStyles,
    ContainerStyles,
    ImageStyles,
    TableStyles,
    WrapperGameStyles,
} from './View.styles.jsx';
import {
    CellType,
    GameMode,
    GameStatus,
    ImageCarousel,
} from '../../common/utils';
import Controls from '../controls/Controller';
import { chooseImage } from './helper/props.js';
import Modal from '../styled-components/components/Modal.jsx';
import Timer from '../timer/View.jsx';

const EndModal = ({ gameStatus, restartGame }) => (
    <Modal
        buttonText='RETRY?'
        handleClick={restartGame}
        image={chooseImage(gameStatus)}
        title={gameStatus.toUpperCase()}
    />
);

const Row = ({ row, lastInput }) => {
    return row.map((cell, index) => <CellStyles key={index}><ImageStyles
        src={ImageCarousel(cell)}
        lastInput={lastInput}
        player={cell === CellType.duna}
    />
    </CellStyles>);
};

const Game = ({
    controls,
    currentPos,
    endGame,
    gameStatus,
    restartGame,
    lastInput,
    matrix,
    mode,
    setCurrentPos,
    setLastInput,
    size,
    time,
}) => {
    return (
        <>
            {matrix ?
                (<WrapperGameStyles>
                    {gameStatus !== GameStatus.playing &&
                        <EndModal gameStatus={gameStatus} restartGame={restartGame} />
                    }

                    {mode === GameMode.trial && <Timer
                            stopTimer={gameStatus !== GameStatus.playing}
                            handleEndOfTimer={() => endGame(GameStatus.defeat)}
                            time={time}
                        />
                    }
                    <ContainerStyles>
                        <TableStyles>
                            {matrix.map((row, index) => <tr key={index}><Row row={row} key={index} lastInput={lastInput} /></tr>)}
                        </TableStyles>
                        <Controls
                            controls={controls}
                            currentPos={currentPos}
                            gameStatus={gameStatus}
                            matrix={matrix}
                            setCurrentPos={setCurrentPos}
                            setLastInput={setLastInput}
                            size={size}
                        />
                    </ContainerStyles>
                </WrapperGameStyles>) :
                null
            }
        </>
    );
}

export default Game;