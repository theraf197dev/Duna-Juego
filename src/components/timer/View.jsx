import React, { useState, useEffect } from 'react';
import { getTimeFormated } from './helper';
import {
    ContainerStyles,
} from './View.styles.jsx';

const Timer = ({
    handleEndOfTimer,
    stopTimer,
    time,
}) => {
  const [seconds, setSeconds] = useState(time);

  useEffect(() => {
    let interval = null;

    if (seconds > 0 && !stopTimer) {
        interval = setInterval(() => {
        setSeconds(seconds => seconds - 1);
      }, 1000);
    } else {
        if (seconds <= 0) {
            handleEndOfTimer();
        }
        clearInterval(interval);
    }

    return () => clearInterval(interval);
  }, [seconds, stopTimer]);

  return (
    <ContainerStyles seconds={seconds}>{getTimeFormated(seconds)}</ContainerStyles>
  );
};

export default Timer;