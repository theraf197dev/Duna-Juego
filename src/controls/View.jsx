import React from 'react';
import {
  ButtonStyles,
  CircleStyles,
  ContainerStyles,
  CrossStyles,
  FigureStyles,
} from './View.styles.jsx';
import { canMove } from '../common/utils.js';

const View = ({
    controls,
    currentPos,
    handleChangePosition,
    matrix,
    size,
}) => {
  return (
    <ContainerStyles>
      {controls.map((control) =>
        <ButtonStyles
          control={control}
          disabled={!canMove(matrix, currentPos, control.vector, control.incr, size)}
          key={control.id}
          onClick={() => handleChangePosition(control.vector, control.incr)}
        >
          <FigureStyles control={control} />
        </ButtonStyles>
      )}
      <CrossStyles>
        <CircleStyles onClick={() => console.log('bark')} />
      </CrossStyles>
    </ContainerStyles>
  )
}

export default View