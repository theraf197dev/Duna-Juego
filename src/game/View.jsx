import React from 'react';
import {
    CellStyles,
    ContainerStyles,
    ImageStyles,
    TableStyles,
} from './View.styles.jsx';
import {
    CellType,
} from '../common/utils';
import {
    ImageCarousel,
} from '../common/utils';
import Controls from '../controls/Controller';

const Row = ({ row, lastInput }) => {
    return row.map((cell, index) => <CellStyles key={index}><ImageStyles src={ImageCarousel(cell)} lastInput={lastInput} player={cell === CellType.duna} /></CellStyles> );
};

const Game = ({
    controls,
    currentPos,
    lastInput,
    matrix,
    setCurrentPos,
    setLastInput,
    size,
}) => {
    return (
        <>
            {matrix ?
                (<>
                    <ContainerStyles>
                        <TableStyles>
                            {matrix.map((row, index) => <tr key={index}><Row row={row} key={index} lastInput={lastInput} /></tr>)}
                        </TableStyles>
                        <Controls
                            controls={controls}
                            currentPos={currentPos}
                            matrix={matrix}
                            setCurrentPos={setCurrentPos}
                            setLastInput={setLastInput}
                            size={size}
                        />
                    </ContainerStyles>
                </>) :
                null
            }
        </>
    );
}

export default Game;