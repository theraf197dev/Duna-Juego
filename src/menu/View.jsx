import React, { useState } from 'react';
import Game from '../game/Model';
import { Toogle, Button } from '../styled-components';
import {
    ContainerStyles,
    GameStyles,
    ImageStyles,
    SpanStyles,
    ToggleStyles,
} from './View.styles.jsx';

const MenuView = ({
    games,
    initGame,
    selected,
    setDifficultyLvl,
    setSelected,
}) => {
    return (
        <form onSubmit={() => initGame()}>
            <ContainerStyles>
                {games.items.map(({name, url}) =>
                    <div key={name}>
                        <h2>{name}</h2>
                        <GameStyles>
                            <ImageStyles src={url} onClick={() => setSelected(name)} />
                            <SpanStyles id={1} selected={selected} name={name} />
                            <SpanStyles id={2} selected={selected} name={name} />
                            <SpanStyles id={3} selected={selected} name={name} />
                            <SpanStyles id={4} selected={selected} name={name} />
                        </GameStyles>
                    </div>
                )}
            </ContainerStyles>
            <ToggleStyles>
                <label>Difficulty</label>
                <Toogle
                    handleChange={(isChecked) => setDifficultyLvl(isChecked + 1)}
                    id='difficulty'
                />
            </ToggleStyles>
            <Button buttonText='Play' />
        </form>
    );
};



const Menu = ({
    board,
    controls,
    difficulty,
    games,
    initGame,
    isGameAlive,
    mode,
    setGame,
    size,
}) => {
    const [selected, setSelected] = useState(mode);
    const [difficultyLvl, setDifficultyLvl] = useState(difficulty);

    return (
        <>
            {isGameAlive ?
                <Game
                    board={board}
                    controls={controls}
                    difficulty={difficultyLvl}
                    mode={selected}
                    setGame={setGame}
                    size={size}
                />
                :
                <MenuView
                    games={games}
                    initGame={initGame}
                    selected={selected}
                    setDifficultyLvl={setDifficultyLvl}
                    setSelected={setSelected}
                />
            }
        </>
    );
}

export default Menu;