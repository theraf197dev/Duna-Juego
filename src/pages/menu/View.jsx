import React, { useState } from 'react';
import Game from '../../components/game/Model';
import { Toogle, Button } from '../../components/styled-components';
import {
    ContainerStyles,
    GameStyles,
    ImageStyles,
    SpanStyles,
    ToggleStyles,
} from './View.styles.jsx';

const MenuView = ({
    difficultyLvl,
    games,
    setGame,
    selected,
    setDifficultyLvl,
    setSelected,
}) => {
    return (
        <>
            <ContainerStyles>
                {games.items.map(({name, url}) =>
                    <div key={name} data-testid='fn-game-item'>
                        <h2>{name}</h2>
                        <GameStyles data-testid='fn-game-item-image'>
                            <ImageStyles
                                onClick={() => setSelected(name)}
                                src={url}
                            />
                            <SpanStyles id={1} selected={selected} name={name} />
                            <SpanStyles id={2} selected={selected} name={name} />
                            <SpanStyles id={3} selected={selected} name={name} />
                            <SpanStyles id={4} selected={selected} name={name} />
                        </GameStyles>
                    </div>
                )}
            </ContainerStyles>
            <ToggleStyles>
                <label>Difficulty</label>
                <Toogle
                    value={difficultyLvl - 1}
                    handleChange={(isChecked) => setDifficultyLvl(isChecked + 1)}
                    id='difficulty'
                />
            </ToggleStyles>
            <Button buttonText='PLAY' handleClick={() => setGame(true)} />
        </>
    );
};

const Menu = ({
    controls,
    difficulty,
    games,
    mode,
    size,
    time,
}) => {
    const [selected, setSelected] = useState(mode);
    const [difficultyLvl, setDifficultyLvl] = useState(difficulty);
    const [isGameAlive, setGame] = useState(false);

    return (
        <>
            {isGameAlive ?
                <Game
                    controls={controls}
                    difficulty={difficultyLvl}
                    mode={selected}
                    restartGame={() => setGame(false)}
                    size={size}
                    time={time}
                />
                :
                <MenuView
                    difficultyLvl={difficultyLvl}
                    games={games}
                    selected={selected}
                    setDifficultyLvl={setDifficultyLvl}
                    setGame={setGame}
                    setSelected={setSelected}
                />
            }
        </>
    );
}

export default Menu;