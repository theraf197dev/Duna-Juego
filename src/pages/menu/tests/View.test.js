import React, { useCallback, useState } from "react";
import { render } from "@testing-library/react";
import { renderHook, act } from "@testing-library/react-hooks";
import View from "../View.jsx";
import fixtures from "../../../common/fixtures.js";

function useHooks() {
  const [difficultyLvl, setDifficultyLvl] = useState(1);

  const changeDifficulty = useCallback(() => setDifficultyLvl(2));

  return { difficultyLvl, changeDifficulty };
};

describe('View tests', () => {
  let props = fixtures;

  it('render component', () => {
      const { container } = render(<View {...props} />);

      expect(container).toMatchSnapshot();
  });

  it('render component without games', () => {
      const testProps = {
        ...props,
        games: {
          items: [],
        },
      };

      const { container } = render(<View {...testProps} />);

      const games = container.querySelector('[data-testid="fn-game-item"]');

      expect(games).not.toBeInTheDocument();
  });

  test('should change difficulty', () => {
    const { result } = renderHook(() => useHooks())

    act(() => {
      result.current.changeDifficulty()
    });

    expect(result.current.difficultyLvl).toBe(2);
  })
});


