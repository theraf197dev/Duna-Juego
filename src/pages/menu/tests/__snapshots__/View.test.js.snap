// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View tests render component 1`] = `
<div>
  <div
    class="sc-dmctIk bhkReJ"
  >
    <div
      data-testid="fn-game-item"
    >
      <h2>
        TRIAL
      </h2>
      <div
        class="sc-bBABsx dUtqQl"
      >
        <img
          class="sc-kgTSHT cOXSoe"
          src="trial_mode.jpg"
        />
        <span
          class="sc-iveFHk hwZLcM"
          id="1"
          name="TRIAL"
        />
        <span
          class="sc-iveFHk ihMfXg"
          id="2"
          name="TRIAL"
        />
        <span
          class="sc-iveFHk dyWYud"
          id="3"
          name="TRIAL"
        />
        <span
          class="sc-iveFHk doamAl"
          id="4"
          name="TRIAL"
        />
      </div>
    </div>
    <div
      data-testid="fn-game-item"
    >
      <h2>
        SEEKER
      </h2>
      <div
        class="sc-bBABsx dUtqQl"
      >
        <img
          class="sc-kgTSHT cOXSoe"
          src="seeker_mode.jpg"
        />
        <span
          class="sc-iveFHk ZNbqd"
          id="1"
          name="SEEKER"
        />
        <span
          class="sc-iveFHk ZNbqd"
          id="2"
          name="SEEKER"
        />
        <span
          class="sc-iveFHk ZNbqd"
          id="3"
          name="SEEKER"
        />
        <span
          class="sc-iveFHk ZNbqd"
          id="4"
          name="SEEKER"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-fLcnxK iGPkiX"
  >
    <label>
      Difficulty
    </label>
    <div
      class="sc-hBxehG fsdLeU"
    >
      <input
        class="sc-fnGiBr hwCGdJ"
        data-testid="fn-toogle-button"
        id="difficulty"
        type="checkbox"
      />
      <label
        class="sc-fEXmlR ccJzPd"
        for="difficulty"
      >
        <div
          class="sc-bjfHbI eqszNL"
        >
          <div
            class="sc-idXgbr gxGckC"
          >
            ★
          </div>
          <div
            class="sc-dIfARi lfcJaa"
          >
            ★
          </div>
        </div>
        <div
          class="sc-hHTYSt kBoZNk"
        />
      </label>
    </div>
  </div>
  <div
    class="sc-eDWCr kzwVnX"
  >
    <button
      class="sc-bqWxrE gTDXbS"
    >
      PLAY
    </button>
  </div>
</div>
`;
